File: service/src/main/java/org/helmut/profile/rest/service/ProjectService.java
package org.helmut.profile.rest.service;

import org.helmut.profile.business.bc.ProjectBC;
import org.helmut.profile.business.model.CreateChapterTO;
import org.helmut.profile.business.model.ProjectTO;
import org.helmut.profile.business.model.SectionTO;
import org.helmut.profile.business.model.UserTO;
import org.helmut.profile.rest.auth.filter.JWTTokenNeeded;

import javax.inject.Inject;
import javax.ws.rs.*;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.HttpHeaders;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.util.List;

import static org.helmut.profile.rest.service.Constants.CURRENT_USER_EMAIL;

@Path("project")
public class ProjectService {

    @Inject
    private ProjectBC projectBC;

    @Context
    private HttpHeaders httpHeaders;

    @GET
    @Path("{name}")
    @Produces(MediaType.APPLICATION_JSON)
    public ProjectTO getByName(@PathParam("name") String name) {
        return projectBC.getByName(name);
    }

    @GET
    @Path("byEmail/{email}")
    @Produces(MediaType.APPLICATION_JSON)
    public List<ProjectTO> getByEmail(@PathParam("email") String email) {
        return projectBC.getByEmail(email);
    }

    @POST
    @JWTTokenNeeded
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    public Response createProject(ProjectTO projectTO) {
        String currentUserEmail = httpHeaders.getHeaderString(CURRENT_USER_EMAIL);
        UserTO userTO = new UserTO();
        userTO.setEmail(currentUserEmail);
        projectTO.setUserTO(userTO);
        try {
            projectBC.createProject(projectTO);
            return Response.ok().build();
        } catch (Exception e) {
            return Response.status(Response.Status.BAD_REQUEST)
                    .entity(e)
                    .build();
        }
    }

    @POST
    @JWTTokenNeeded
    @Path("chapter")
    @Consumes(MediaType.APPLICATION_JSON)
    public Response createChapter(CreateChapterTO createChapterTO) {
        String currentUserEmail = httpHeaders.getHeaderString(CURRENT_USER_EMAIL);
        try {
            projectBC.createChapter(createChapterTO, currentUserEmail);
            return Response.ok().build();
        } catch (Exception e) {
            return Response.status(Response.Status.BAD_REQUEST)
                    .entity(e)
                    .build();
        }
    }

    @PUT
    @JWTTokenNeeded
    @Path("section")
    @Consumes(MediaType.APPLICATION_JSON)
    public Response updateSection(SectionTO sectionTO) {
        String currentUserEmail = httpHeaders.getHeaderString(CURRENT_USER_EMAIL);
        try {
            projectBC.updateSection(sectionTO, currentUserEmail);
            return Response.ok().build();
        } catch (IllegalArgumentException e) {
            return Response.status(Response.Status.BAD_REQUEST)
                    .entity(e)
                    .build();
        }
    }
}

File: web-client/src/main/js/profile/src/app/projects/service/project.service.ts
import {Injectable} from '@angular/core';
import {Observable} from "rxjs";
import {ProjectTO} from "../../shared/model/to/project-t-o";
import {BackendService} from "../../shared/service/backend.service";
import {CreateChapterTO} from "../../shared/model/to/create-chapter-t-o";
import {SectionTO} from "../../shared/model/to/section-t-o";

@Injectable({
  providedIn: 'root'
})
export class ProjectService {

  constructor(private backend: BackendService) {
  }

  public getProjectByEmail(email: string): Observable<ProjectTO[]> {
    return this.backend.get<ProjectTO[]>('project/byEmail/' + email);
  }

  public getByName(name: string): Observable<ProjectTO> {
    return this.backend.get<ProjectTO>('project/' + name);
  }

  public createProject(projectTO: ProjectTO): Observable<any> {
    return this.backend.post('project', projectTO);
  }

  public createChapter(createChapterTO: CreateChapterTO): Observable<any> {
    return this.backend.post('project/chapter', createChapterTO);
  }

  public updateSection(sectionTO: SectionTO): Observable<any> {
    return this.backend.put('project/section', sectionTO);
  }
}
